cpp/src/arrow/python/numpy_to_arrow.cc

```diff
@@ -424,38 +451,12 @@ inline Status NumPyConverter::ConvertData(std::shared_ptr<Buffer>* data) {
   return Status::OK();
 }
 
-template <>
-inline Status NumPyConverter::ConvertData<BooleanType>(std::shared_ptr<Buffer>* data) {
-  int64_t nbytes = BitUtil::BytesForBits(length_);
-  std::shared_ptr<Buffer> buffer;
-  RETURN_NOT_OK(AllocateBuffer(pool_, nbytes, &buffer));
-
-  Ndarray1DIndexer<uint8_t> values(arr_);
```

Comment:
Seems like a dtype check did not happen, why was this code path hitting silently? I can take a closer look too but curious if you know

Reply:
The output array is visited and there is no check for the dtype except in the spec`ConvertData` specialized for date types.  In the case of a boolean output type, it assumed the numpy data was `unit8_t`.  In the other cases, the dtype is sent to the `CastBuffer` where the input and output types are parameterized, but it expects the buffer to be in Arrow layout, so for bools it needs to be converted to a bitmask before.